library/core/src/io/borrowed_buf.rs

```diff
@@ -353,4 +370,32 @@ impl<'a> BorrowedCursor<'a> {
         }
         self.buf.filled += buf.len();
     }
+
+    /// Runs the given closure with a `BorrowedBuf` containing the unfilled part
+    /// of the cursor.
+    ///
+    /// This enables inspecting what was written to the cursor.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the `BorrowedBuf` given to the closure is replaced by another
+    /// one.
+    pub fn with_unfilled_buf<T>(&mut self, f: impl FnOnce(&mut BorrowedBuf<'_>) -> T) -> T {
+        let mut buf = BorrowedBuf::from(self.reborrow());
+        let prev_ptr = buf.buf as *const _;
+        let res = f(&mut buf);
+
+        // Check that the caller didn't replace the `BorrowedBuf`.
+        assert!(core::ptr::addr_eq(prev_ptr, buf.buf));
```

Comment:
Is this a necessary condition for soundness? If so, we probably have to abort here, not just panic. Can you clarify in comments why it's necessary to check this as well?

(Also if it's a soundness condition it should probably be in a drop guard, so that the closure panicking after replacing the Buf doesn't cause us to unwind past this check).